### ISSUE TITLE
scheduler takes a long time to shutdown

### ISSUE THREAD

**Issue by sharnoff** [neondatabase/autoscaling#55]
Thought it'd be fixed by #32, but seems like it wasn't.

At first glance, seems like the context solution doesn't work either way, because K8s doesn't use the cobra command's context - see:

> ```go
> func runCommand(cmd *cobra.Command, opts *options.Options, registryOptions ...Option) error {
> 	verflag.PrintAndExitIfRequested()
> 
> 	// Activate logging as soon as possible, after that
> 	// show flags with the final logging configuration.
> 	if err := logsapi.ValidateAndApply(opts.Logs, utilfeature.DefaultFeatureGate); err != nil {
> 		fmt.Fprintf(os.Stderr, "%v\n", err)
> 		os.Exit(1)
> 	}
> 	cliflag.PrintFlags(cmd.Flags())
> 
> 	ctx, cancel := context.WithCancel(context.Background())
> 	//                                ^^^^^^^^^^^^^^^^^^^^ should be cmd.Context()
> 	defer cancel()
> 	go func() {
> 		stopCh := server.SetupSignalHandler()
> 		<-stopCh
> 		cancel()
> 	}()
> 
> 	cc, sched, err := Setup(ctx, opts, registryOptions...)
> 	if err != nil {
> 		return err
> 	}
> 	// add feature enablement metrics
> 	utilfeature.DefaultMutableFeatureGate.AddMetrics()
> 	return Run(ctx, cc, sched)
> 	//         ^^^ context used here.
> }
> ```

pointing comments added by me. source: https://github.com/kubernetes/kubernetes/blob/d6fe718e19881399fbb65f693c0c0f29f8f63861/cmd/kube-scheduler/app/server.go#L121-L147

---

HOWEVER, we're working with K8s 1.23 and I don't particularly want to fork kubernetes to backport a fix. AND this shouldn't matter — K8s is using `server.SetupSignalHandler()` to cancel the context, which _should_ work anyways.

**Comment by sharnoff** [comment]
Checked, this issue still occurs after switching to 1.24 — will investigate further later this week.

**Comment by sharnoff** [comment]
Looked into this, tried the fix in this issue, didn't fix it. Going to try the long way (start with base scheduler, gradually add each piece of the plugin until the behavior re-emerges).

**Comment by sharnoff** [comment]
The issue was fixed in later K8s versions by this PR: https://github.com/kubernetes/kubernetes/pull/110207. Specifically, this diff:

```diff
diff --git a/pkg/scheduler/scheduler.go b/pkg/scheduler/scheduler.go
index 8d9cc2c2763..c165b34ffc2 100644
--- a/pkg/scheduler/scheduler.go
+++ b/pkg/scheduler/scheduler.go
@@ -335,7 +335,16 @@ func New(client clientset.Interface,
 // Run begins watching and scheduling. It starts scheduling and blocked until the context is done.
 func (sched *Scheduler) Run(ctx context.Context) {
        sched.SchedulingQueue.Run()
-       wait.UntilWithContext(ctx, sched.scheduleOne, 0)
+
+       // We need to start scheduleOne loop in a dedicated goroutine,
+       // because scheduleOne function hangs on getting the next item
+       // from the SchedulingQueue.
+       // If there are no new pods to schedule, it will be hanging there
+       // and if done in this goroutine it will be blocking closing
+       // SchedulingQueue, in effect causing a deadlock on shutdown.
+       go wait.UntilWithContext(ctx, sched.scheduleOne, 0)
+
+       <-ctx.Done()
        sched.SchedulingQueue.Close()
 }
```

Of course, only discovered this _after_ figuring out the fix myself and then finding it had been fixed in master. :disappointed:

Unfortunately, it wasn't backported to 1.24. It _is_ included in 1.25, but upgrading to 1.25 involves more infra work.
